q2cli/builtin/tools.py

```diff
@@ -195,7 +196,111 @@ def peek(path):
+@click.argument('paths', nargs=-1, required=True, metavar='METADATA...',
+                type=click.Path(exists=True, file_okay=True, dir_okay=False,
+                                readable=True))
+def cast_metadata(paths, cast, output_file, ignore_extra,
+                  error_on_missing):
+    import tempfile
+    import qiime2
+    import sys
+
+    metadata = _merge_metadata(paths)
+
+    cast_dict = {}
+    try:
+        for casting in cast:
+            col, type = casting.split(':')
+            if col in cast_dict:
+                raise click.BadParameter(
+                    message=(f'Column name "{col}" appears in cast more than'
+                             ' once.'),
+                    param_hint='cast')
+            cast_dict[col] = type
+    except Exception as err:
+        header = \
+            ('Could not parse provided cast arguments into unique COLUMN:TYPE'
+             ' pairs. Please make sure all cast flags are of the format --cast'
+             ' COLUMN:TYPE')
+        q2cli.util.exit_with_error(err, header=header)
+
+    types = set(cast_dict.values())
+    if not types.issubset(_COLUMN_TYPES):
+        raise click.BadParameter(
+            message=('Unknown column type provided. Please make sure all'
+                     ' columns included in your cast contain a valid column'
+                     ' type. Valid types: %s' %
+                     (sorted(_COLUMN_TYPES))),
+            param_hint='cast')
+
+    column_names = set(metadata.columns.keys())
+    cast_names = set(cast_dict.keys())
+
+    if not ignore_extra:
+        if not cast_names.issubset(column_names):
+            cast = cast_names.difference(column_names)
+            raise click.BadParameter(
+                message=('The following cast columns were not found'
+                         ' within the metadata: %s' %
+                         (sorted(cast))),
+                param_hint='cast')
+
+    if error_on_missing:
+        if not column_names.issubset(cast_names):
+            cols = column_names.difference(cast_names)
+            raise click.BadParameter(
+                message='The following columns within the metadata'
+                        ' were not provided in the cast: %s' %
+                        (sorted(cols)),
+                param_hint='cast')
+
+    # Remove entries from the cast dict that are not in the metadata to avoid
+    # errors further down the road
+    for cast in cast_names:
+        if cast not in column_names:
+            cast_dict.pop(cast)
+
+    with tempfile.NamedTemporaryFile() as temp:
+        metadata.save(temp.name)
+        cast_md = qiime2.Metadata.load(temp.name, cast_dict)
+
+    if output_file:
+        cast_md.save(output_file)
+    else:
+        with tempfile.NamedTemporaryFile(mode='w+') as stdout_temp:
+            cast_md.save(stdout_temp.name)
+            stdout_str = stdout_temp.read()
+            sys.stdout.write(stdout_str)
```

Comment:
Ahhh yes, good catch! I think in this repo we primarily use a click utility (`click.echo`) for handling this kind of stream writing:

https://github.com/qiime2/q2cli/search?q=click.echo

Can you update this to make it consistent with the rest of the package? Thanks!